File: ts/examples/stock-forecast.ts
/**
 * Example: Fetch stock data and forecast prices
 *
 * Run with:
 *   npx ts-node examples/stock-forecast.ts
 *
 * Requires: npm install yahoo-finance2
 */

import {
  initWasm,
  Arima,
  Holt,
  LinearRegression,
  mae,
  rmse,
  mape,
  trainTestSplit,
} from '../src';

// Simple Yahoo Finance fetcher using the public API
async function fetchYahooData(
  symbol: string,
  startDate: string,
  endDate: string
): Promise<{ dates: string[]; prices: number[] }> {
  const start = Math.floor(new Date(startDate).getTime() / 1000);
  const end = Math.floor(new Date(endDate).getTime() / 1000);

  const url = `https://query1.finance.yahoo.com/v8/finance/chart/${symbol}?period1=${start}&period2=${end}&interval=1d`;

  const response = await fetch(url);
  const data = await response.json();

  if (data.chart.error) {
    throw new Error(data.chart.error.description);
  }

  const result = data.chart.result[0];
  const timestamps: number[] = result.timestamp;
  const closes: number[] = result.indicators.quote[0].close;

  const dates: string[] = [];
  const prices: number[] = [];

  for (let i = 0; i < timestamps.length; i++) {
    if (closes[i] != null) {
      dates.push(new Date(timestamps[i] * 1000).toISOString().split('T')[0]);
      prices.push(closes[i]);
    }
  }

  return { dates, prices };
}

async function main() {
  console.log('=== Stock Price Forecasting Demo ===\n');

  // Initialize WASM
  await initWasm();

  // Fetch data
  const symbol = 'AAPL';
  const startDate = '2024-01-01';
  const endDate = '2024-12-01';

  console.log(`Fetching ${symbol} data from ${startDate} to ${endDate}...`);

  const { dates, prices } = await fetchYahooData(symbol, startDate, endDate);
  console.log(`Retrieved ${prices.length} data points\n`);

  // Split train/test
  const { train, test } = trainTestSplit(prices, 0.2);
  const forecastHorizon = test.length;

  console.log(`Training set: ${train.length} points`);
  console.log(`Test set: ${test.length} points\n`);

  // Price stats
  const minPrice = Math.min(...prices);
  const maxPrice = Math.max(...prices);
  console.log(`Price range: $${minPrice.toFixed(2)} - $${maxPrice.toFixed(2)}\n`);

  console.log('=== Model Comparison ===\n');

  // 1. Linear Regression
  const linear = new LinearRegression();
  await linear.fit(train);
  const linearForecast = await linear.predict(forecastHorizon);

  console.log('Linear Regression:');
  console.log(`  Slope: ${linear.getSlope().toFixed(4)} ($ per day)`);
  console.log(`  R²: ${linear.getRSquared().toFixed(4)}`);
  printMetrics(test, linearForecast);

  // 2. Holt
  const holt = new Holt(0.3, 0.1);
  await holt.fit(train);
  const holtForecast = await holt.predict(forecastHorizon);

  console.log('Holt (Double Exp. Smoothing):');
  printMetrics(test, holtForecast);

  // 3. ARIMA
  const arima = new Arima(1, 1, 0);
  await arima.fit(train);
  const arimaForecast = await arima.predict(forecastHorizon);

  console.log('ARIMA(1,1,0):');
  printMetrics(test, arimaForecast);

  // Sample predictions
  console.log('\n=== Sample Predictions (last 5 days) ===\n');
  console.log('Date        |    Actual |     ARIMA');
  console.log('------------|-----------|----------');

  const startIdx = Math.max(0, test.length - 5);
  for (let i = startIdx; i < test.length; i++) {
    const dateIdx = train.length + i;
    const date = dates[dateIdx] || `Day ${dateIdx + 1}`;
    console.log(
      `${date} | $${test[i].toFixed(2).padStart(8)} | $${arimaForecast[i].toFixed(2).padStart(8)}`
    );
  }

  // Returns analysis
  console.log('\n=== Returns Analysis ===\n');
  const returns = prices.slice(1).map((p, i) => (p - prices[i]) / prices[i]);
  const avgReturn = returns.reduce((a, b) => a + b, 0) / returns.length;
  const volatility = Math.sqrt(
    returns.reduce((acc, r) => acc + (r - avgReturn) ** 2, 0) / returns.length
  );

  console.log(`Average daily return: ${(avgReturn * 100).toFixed(4)}%`);
  console.log(`Daily volatility: ${(volatility * 100).toFixed(4)}%`);
  console.log(`Annualized volatility: ${(volatility * Math.sqrt(252) * 100).toFixed(2)}%`);
}

function printMetrics(actual: number[], predicted: number[]) {
  console.log(`  MAE:  $${mae(actual, predicted).toFixed(2)}`);
  console.log(`  RMSE: $${rmse(actual, predicted).toFixed(2)}`);
  console.log(`  MAPE: ${(mape(actual, predicted) * 100).toFixed(2)}%\n`);
}

main().catch(console.error);

File: examples/stock_forecast.rs
//! Example: Fetch stock data and forecast prices
//!
//! Run with:
//! ```bash
//! cargo run --example stock_forecast --features fetch
//! ```

use rustful_ts::algorithms::{
    arima::Arima,
    exponential_smoothing::DoubleExponentialSmoothing,
    linear_regression::LinearRegression,
    Predictor,
};
use rustful_ts::data::{fetch_stock_sync, closing_prices, daily_returns, Interval};
use rustful_ts::utils::metrics::{mae, rmse, mape};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("=== Stock Price Forecasting Demo ===\n");

    // Fetch Apple stock data
    let symbol = "AAPL";
    let start_date = "2024-01-01";
    let end_date = "2024-12-01";

    println!("Fetching {} data from {} to {}...", symbol, start_date, end_date);

    let quotes = fetch_stock_sync(symbol, start_date, end_date, Interval::Daily)?;
    println!("Retrieved {} data points\n", quotes.len());

    // Extract closing prices
    let prices = closing_prices(&quotes);

    // Split into train/test (80/20)
    let split_idx = (prices.len() as f64 * 0.8) as usize;
    let train = &prices[..split_idx];
    let test = &prices[split_idx..];
    let forecast_horizon = test.len();

    println!("Training set: {} points", train.len());
    println!("Test set: {} points\n", test.len());

    // Show price range
    let min_price = prices.iter().cloned().fold(f64::INFINITY, f64::min);
    let max_price = prices.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
    println!("Price range: ${:.2} - ${:.2}\n", min_price, max_price);

    // Test different models
    println!("=== Model Comparison ===\n");

    // 1. Linear Regression
    {
        let mut model = LinearRegression::new();
        model.fit(train)?;
        let forecast = model.predict(forecast_horizon)?;

        println!("Linear Regression:");
        println!("  Slope: {:.4} ($ per day)", model.slope());
        println!("  R²: {:.4}", model.r_squared());
        print_metrics(test, &forecast);
    }

    // 2. Holt (Double Exponential Smoothing)
    {
        let mut model = DoubleExponentialSmoothing::new(0.3, 0.1)?;
        model.fit(train)?;
        let forecast = model.predict(forecast_horizon)?;

        println!("Holt (Double Exp. Smoothing):");
        print_metrics(test, &forecast);
    }

    // 3. ARIMA
    {
        let mut model = Arima::new(1, 1, 0)?;
        model.fit(train)?;
        let forecast = model.predict(forecast_horizon)?;

        println!("ARIMA(1,1,0):");
        print_metrics(test, &forecast);
    }

    // 4. ARIMA with higher order
    {
        let mut model = Arima::new(2, 1, 1)?;
        model.fit(train)?;
        let forecast = model.predict(forecast_horizon)?;

        println!("ARIMA(2,1,1):");
        print_metrics(test, &forecast);
    }

    // Show sample predictions
    println!("\n=== Sample Predictions (last 5 days) ===\n");
    println!("{:<12} {:>12} {:>12}", "Date", "Actual", "ARIMA");

    let mut model = Arima::new(1, 1, 0)?;
    model.fit(train)?;
    let forecast = model.predict(forecast_horizon)?;

    for i in (test.len().saturating_sub(5))..test.len() {
        println!(
            "{:<12} ${:>10.2} ${:>10.2}",
            format!("Day {}", split_idx + i + 1),
            test[i],
            forecast[i]
        );
    }

    // Daily returns analysis
    println!("\n=== Returns Analysis ===\n");
    let returns = daily_returns(&prices);
    let avg_return = returns.iter().sum::<f64>() / returns.len() as f64;
    let volatility = (returns.iter().map(|r| (r - avg_return).powi(2)).sum::<f64>()
        / returns.len() as f64)
        .sqrt();

    println!("Average daily return: {:.4}%", avg_return * 100.0);
    println!("Daily volatility: {:.4}%", volatility * 100.0);
    println!(
        "Annualized volatility: {:.2}%",
        volatility * (252.0_f64).sqrt() * 100.0
    );

    Ok(())
}

fn print_metrics(actual: &[f64], predicted: &[f64]) {
    println!("  MAE:  ${:.2}", mae(actual, predicted));
    println!("  RMSE: ${:.2}", rmse(actual, predicted));
    println!("  MAPE: {:.2}%\n", mape(actual, predicted) * 100.0);
}
